### DISCUSSION TITLE
cloud-hypervisor CLI support create and boot a virtual machine by a config file

### DISCUSSION THREAD

**Discussion by scuzhanglei** [cloud-hypervisor/cloud-hypervisor#4238] [Ideas]
Now CLI only suport Create and boot a complete virtual machine by using the CLI with separated args,  it would be nice to also support this by a vmConfig file instead of args.

**Comment by sboeuf** [comment]
Ok so just to clarify, do you mean you'd like to start a VM with something like:
```bash
cloud-hypervisor --config-file /foo/bar/my_vm_config_file
```
?

**Reply by scuzhanglei** [reply]
yes, it is.

**Reply by sboeuf** [reply]
@rbradford @liuw @michael2012z what do you think about this? Is that something that would be acceptable for the project? Do you see a need for this?

**Reply by liuw** [reply]
Having a parser for a bespoke format is a maintenance burden.

Why not accept a serialized format for the `VmConfig` type and call it a day?

**Reply by michael2012z** [reply]
This requirement reminds me the XML configuration of `libvirt`: https://libvirt.org/formatdomain.html
Maybe the support of using a configuration file should be achieved via `libvirt`. That way, the file format is standardized, and CH itself is free from parsing files.

**Reply by rbradford** [reply]
> @rbradford @liuw @michael2012z what do you think about this? Is that something that would be acceptable for the project? Do you see a need for this?

Two approaches:

1. We could add a new parameter to `cloud-hypervisor` itself but that would complicate the logic about whether to start in API server only mode (no VM)
2. We add a create command to `ch-remote` that takes the JSON `VmConfig` config from stdin or from a file.

Option 2 would be my preference.

**Reply by liuw** [reply]
> This requirement reminds me the XML configuration of `libvirt`: https://libvirt.org/formatdomain.html Maybe the support of using a configuration file should be achieved via `libvirt`. That way, the file format is standardized, and CH itself is free from parsing files.

That will force people to use libvirt. I don't think that's desirable for many use cases.

**Reply by sboeuf** [reply]
> > @rbradford @liuw @michael2012z what do you think about this? Is that something that would be acceptable for the project? Do you see a need for this?
> 
> Two approaches:
> 
>     1. We could add a new parameter to `cloud-hypervisor` itself but that would complicate the logic about whether to start in API server only mode (no VM)
> 
>     2. We add a create command to `ch-remote` that takes the JSON `VmConfig` config from stdin or from a file.
> 
> 
> Option 2 would be my preference.

Yes option 2 is the less intrusive one :+1:
